addons/account/models/account_tax.py

```diff
@@ -2394,8 +2397,10 @@ def _prepare_tax_lines(self, base_lines, company, tax_lines=None):
             k: v
             for k, v in tax_lines_mapping.items()
             if (
-                k.get('currency_id') and not self.env['res.currency'].browse(k['currency_id']).is_zero(v['amount_currency'])
-                or not company.currency_id.is_zero(v['balance'])
+                self._should_keep_zero_lines(base_lines and base_lines[0].get('record')) or (
+                    k.get('currency_id') and not self.env['res.currency'].browse(k['currency_id']).is_zero(v['amount_currency'])
+                    or not company.currency_id.is_zero(v['balance'])
+                )
```

Comment:
You no longer have the base lines here. They are already aggregated. Look above:
```py
                    grouping_key = frozendict(tax_rep_data['grouping_key'])
                    tax_line = tax_lines_mapping[grouping_key]
```
Add a value in the grouping key instead. That way, you will have access to base_line / tax_data / tax_rep_data.
Then,
```suggestion
                v['keep_zero_line'] or (
                    k.get('currency_id') and not self.env['res.currency'].browse(k['currency_id']).is_zero(v['amount_currency'])
                    or not company.currency_id.is_zero(v['balance'])
                )
```

Reply:
Hello, thanks for checking! Whether taxes are calculated externally is based on the containing record (e.g., `account.move` or `sale.order`). AFAICT all `base_lines` in `_prepare_tax_lines` always belong to the same containing record. So that's why I could get away with just looking at the first base line. But it's not very nice, your solution is definitely better.

I immediately do `v.pop('keep_zero_line')` so it doesn't end up in the `create` of `account.move.line`. Not sure how this will be used by other countries, but I added all three parameters to the method so the maximum amount of info is there for those who need it.